### PR TITLE
add im revision support

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
In latest spec, we have added Interaction model revision as common tag for every action

#### Change overview
Add InteractionModel Revision tag in last one, 0xFF, for each IM message.
Add MessageParser, MessageBuilder to hold on the shared logic for encode/decode of InteractionModelRevision.
We add it when closing the message's container.
Fix-up the reserved buffer logic across.

#### Testing
The existing test covers

### CODE REVIEW

File: src/app/MessageDef/MessageDefHelper.cpp [hunk 1]
```diff
@@ -94,5 +95,33 @@ void DecreaseDepth()
...
 #endif
+
+#if CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
+CHIP_ERROR CheckInteractionModelRevision(int & aTagPresenceMask, uint8_t aTag, TLV::TLVReader & aReader)
+{
+    VerifyOrReturnError(!(aTagPresenceMask & 1 << aTag), CHIP_ERROR_INVALID_TLV_TAG);
+    aTagPresenceMask |= (1 << aTag);
+#if CHIP_DETAIL_LOGGING
+    {
+        uint8_t interactionModelRevision = 0;
```
Comment (Damian-Nordic) [inline src/app/MessageDef/MessageDefHelper.cpp:106]: Should this value be checked against the local revision or it's just informative?

File: src/include/platform/CHIPDeviceConfig.h [hunk 2]
```diff
@@ -228,6 +228,15 @@
 #define CHIP_DEVICE_CONFIG_DEVICE_SOFTWARE_VERSION 0
 #endif
 
+/**
+ * CHIP_DEVICE_CONFIG_DEVICE_INTERACTION_MODEL_REVISION
+ *
+ * A monothonic number identifying the interaction model revision running on the device.
+ */
+#ifndef CHIP_DEVICE_CONFIG_DEVICE_INTERACTION_MODEL_REVISION
+#define CHIP_DEVICE_CONFIG_DEVICE_INTERACTION_MODEL_REVISION 1
```
Comment (Damian-Nordic) [inline src/include/platform/CHIPDeviceConfig.h:237]: Should this be part of the `CHIPDeviceConfig.h`? I think it usually contains settings which may different across platforms while the IM revision is probably related to the implementation in `src/app`, right?

**Review by Damian-Nordic — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

File: src/app/MessageDef/MessageDefHelper.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/MessageDef/MessageDefHelper.cpp:106]: it is just informative here, thanks

File: src/include/platform/CHIPDeviceConfig.h (hunk 2 above)
Comment (yunhanw-google) [reply src/include/platform/CHIPDeviceConfig.h:237]: @bzbarsky-apple ?

Comment (yunhanw-google) [reply src/include/platform/CHIPDeviceConfig.h:237]: currently I follow the pattern, like software version in ConfigurationManager to do IM revision, it seems we can only put one macro in src/app/InteractionModelRevision.h without integration with ConfigurationManager, which is enough @bzbarsky-apple

Comment (bzbarsky-apple) [reply src/include/platform/CHIPDeviceConfig.h:237]: I see no reason for this to be configurable other than by doing a pull from the SDK repo, so no need for this to be in the overridable device config or configuration manager.

Comment (yunhanw-google) [reply src/include/platform/CHIPDeviceConfig.h:237]: ok, would update it now, thanks

Comment (yunhanw-google) [reply src/include/platform/CHIPDeviceConfig.h]: updated, thanks

**Review by jmartinez-silabs — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/MessageDef/ReadRequestMessage.h [hunk 3]
```diff
@@ -34,11 +35,12 @@ namespace app {
...
-    kDataVersionFilters = 1,
-    kEventRequests      = 2,
-    kEventFilters       = 3,
-    kIsFabricFiltered   = 4,
+    kAttributeRequests        = 0,
+    kDataVersionFilters       = 1,
+    kEventRequests            = 2,
+    kEventFilters             = 3,
+    kIsFabricFiltered         = 4,
+    kInteractionModelRevision = 5,
```
Comment (mrjerryjohns) [inline src/app/MessageDef/ReadRequestMessage.h:43]: Didn't we agree to have the tag value for this be at the end of the tag space? (i.e 0xFF)

More-over, I'd rather such a tag be defined once in a central place, and referenced by all.

File: src/app/TimedRequest.cpp [hunk 4]
```diff
@@ -33,10 +33,10 @@ using namespace Messaging;
...
 {
     // The payload is an anonymous struct (2 bytes) containing a single
-    // 16-bit integer with a context tag (1 control byte, 1 byte tag, at
-    // most 2 bytes for the integer).  Use MessagePacketBuffer::New to
-    // account for other message-global overheads (MIC, etc).
-    System::PacketBufferHandle payload = MessagePacketBuffer::New(6);
+    // 16-bit integer with two context tag (1 control byte, 1 byte tag, at
+    // most 2 bytes for the timeout integer and 1 control byte, 1 byte tag, one byte for InteractionModelRevision).  Use
+    // MessagePacketBuffer::New to account for other message-global overheads (MIC, etc).
+    System::PacketBufferHandle payload = MessagePacketBuffer::New(9);
```
Comment (mrjerryjohns) [inline src/app/TimedRequest.cpp:39]: So, I don't think that the revision tag should actually be encoded right at the end. That results in all kinds of fix-ups to these heuristics, which are not necessary.

Comment (yunhanw-google) [reply src/app/TimedRequest.cpp:39]: Currently only TimedRequest have this issue, I think it is fine.

File: src/app/MessageDef/ReadRequestMessage.h (hunk 3 above)
Comment (yunhanw-google) [reply src/app/MessageDef/ReadRequestMessage.h]: yes, update it with 0xFF, yes, update it using central place(MessageParser, MessageBuilder) to hold shared logic, thanks

**Pull request merged**